spark/src/main/scala/org/apache/spark/sql/delta/perf/OptimizeMetadataOnlyDeltaQuery.scala

```diff
@@ -16,81 +16,358 @@
-          PhysicalOperation(_, Nil, DeltaTable(i: TahoeLogFileIndex))) if i.partitionFilters.isEmpty
-        => extractGlobalCount(i)
-      case _ => None
+  private def extractMinMaxFromPartitionValue(
+      snapshot: Snapshot,
+      lowerCaseColumnNames: Set[String]): Map[String, DeltaColumnStat] = {
+
+    val partitionedColumns = snapshot.metadata.partitionSchema
+      .filter(col => lowerCaseColumnNames.contains(col.name.toLowerCase(Locale.ROOT)))
+      .map(col => (col, DeltaColumnMapping.getPhysicalName(col)))
+
+    if (partitionedColumns.isEmpty) {
+      Map.empty
+    } else {
+      val partitionedColumnsValues = partitionedColumns.map { partitionedColumn =>
+        val physicalName = partitionedColumn._2
+        col(s"partitionValues.`$physicalName`")
+          .cast(partitionedColumn._1.dataType).as(physicalName)
+      }
+
+      val partitionedColumnsAgg = partitionedColumns.flatMap { partitionedColumn =>
+        val physicalName = partitionedColumn._2
+
+        Seq(min(s"`$physicalName`").as(s"min_$physicalName"),
+          max(s"`$physicalName`").as(s"max_$physicalName"))
+      }
+
+      val partitionedColumnsQuery = snapshot.allFiles
+        .select(partitionedColumnsValues: _*)
+        .agg(partitionedColumnsAgg.head, partitionedColumnsAgg.tail: _*)
+        .head()
+
+      partitionedColumns.map { partitionedColumn =>
+        val physicalName = partitionedColumn._2
+
+        partitionedColumn._1.name ->
+          DeltaColumnStat(
+            partitionedColumnsQuery.getAs(s"min_$physicalName"),
+            partitionedColumnsQuery.getAs(s"max_$physicalName"))
+      }.toMap
     }
   }
 
-  /** e.g. sql("SELECT COUNT(*) FROM <delta-table>").show() */
-  object ShowCountStarDeltaTable {
-    def unapply(plan: Aggregate): Option[UTF8String] = plan match {
+  /**
+  * Extract the Count, Min and Max values from Delta Log stats and partitionValues.
+  * The first field is the rows count in the table or `None` if we cannot calculate it from stats
+  * If the column is not partitioned, the values are extracted from stats when it exists.
+  * If the column is partitioned, the values are extracted from partitionValues.
+  */
+  private def extractCountMinMaxFromDeltaLog(
+      tahoeLogFileIndex: TahoeLogFileIndex,
+      lowerCaseColumnNames: Set[String]):
+  (Option[Long], CaseInsensitiveMap[DeltaColumnStat]) = {
+    val deltaScanGen = getDeltaScanGenerator(tahoeLogFileIndex)
+
+    val partitionedValues = extractMinMaxFromPartitionValue(
+      deltaScanGen.snapshotToScan,
+      lowerCaseColumnNames)
+
+    val partitionedColNames = partitionedValues.keySet.map(_.toLowerCase(Locale.ROOT))
+    val dataColumnNames = lowerCaseColumnNames -- partitionedColNames
+    val (rowCount, columnStats) = extractCountMinMaxFromStats(deltaScanGen, dataColumnNames)
+
+    (rowCount, CaseInsensitiveMap(columnStats ++ partitionedValues))
+  }
+
+  object AggregateDeltaTable {
```

Comment:
Nit: How about s/AggregateDeltaTable/MetadataOptimizableAggregate? The name may be a bit verbose, but it makes the object's intention very clear.

Reply:
Correct, the method createLocalRelationPlan is only called if query matches AggregateDeltaTable, where we check these conditions